File: src/Azdo.Tui/Components/ListPicker.cs
using Azdo.Tui.Rendering;
using Azdo.Tui.Runtime;
using StyleSet = Azdo.Tui.Styles.Styles;

namespace Azdo.Tui.Components;

/// <summary>A single option in a <see cref="ListPicker"/> (≈ <c>components.ListPickerOption</c>).</summary>
public sealed class ListPickerOption
{
    public string Name { get; set; } = "";
    public string Icon { get; set; } = "";
    public bool IsCurrent { get; set; }
}

/// <summary>Emitted when a list picker option is chosen. An empty value means "clear filter".</summary>
public sealed record ListPickerSelectedMsg(string Value) : IMsg;

/// <summary>
/// A generic single-select modal list picker (≈ <c>components.ListPicker</c>).
/// Used directly and as the model for the more specialized pickers.
/// </summary>
public sealed class ListPicker
{
    internal const int MinModalWidth = 80;

    private readonly StyleSet _styles;
    private bool _visible;
    private int _width;
    private int _height;
    private string _title = "";
    private List<ListPickerOption> _options = new();
    private int _cursor;
    private bool _allowClear;

    public ListPicker(StyleSet styles) => _styles = styles;

    /// <summary>
    /// Configures the picker. When <paramref name="activeValue"/> is set and
    /// <paramref name="allowClear"/> is true, a "Clear filter" option is prepended
    /// and the cursor is positioned on the active value.
    /// </summary>
    public void SetConfig(string title, IEnumerable<ListPickerOption> options, string activeValue, bool allowClear)
    {
        _title = title;
        _allowClear = allowClear;
        _options = new List<ListPickerOption>();
        _cursor = 0;

        bool hasActive = activeValue != "" && allowClear;

        if (hasActive)
            _options.Add(new ListPickerOption { Name = "Clear filter", Icon = "✕", IsCurrent = false });

        int offset = hasActive ? 1 : 0;

        var src = options.ToList();
        for (int i = 0; i < src.Count; i++)
        {
            var opt = src[i];
            opt.IsCurrent = opt.Name == activeValue;
            _options.Add(opt);
            if (opt.Name == activeValue)
                _cursor = i + offset;
        }
    }

    public void Show() => _visible = true;
    public void Hide() => _visible = false;
    public bool IsVisible => _visible;
    public void SetSize(int width, int height) { _width = width; _height = height; }
    public int GetCursor() => _cursor;

    public Cmd? Update(IMsg msg)
    {
        if (!_visible) return null;
        if (msg is KeyMsg key)
        {
            switch (key.Key)
            {
                case "esc":
                case "q":
                    _visible = false;
                    return null;
                case "up":
                case "k":
                    if (_cursor > 0) _cursor--;
                    return null;
                case "down":
                case "j":
                    if (_cursor < _options.Count - 1) _cursor++;
                    return null;
                case "enter":
                    if (_options.Count == 0) return null;
                    var selected = _options[_cursor];
                    _visible = false;
                    string value = selected.Name == "Clear filter" ? "" : selected.Name;
                    return Commands.Of(new ListPickerSelectedMsg(value));
            }
        }
        return null;
    }

    public string View()
    {
        if (!_visible) return "";

        var theme = _styles.Theme;
        const string helpTextStr = "↑/↓: navigate • enter: select • esc/q: cancel";

        int maxWidth = MinModalWidth;
        if (_title.Length > maxWidth) maxWidth = _title.Length;
        if (helpTextStr.Length > maxWidth) maxWidth = helpTextStr.Length;

        foreach (var opt in _options)
        {
            string label = opt.IsCurrent ? opt.Name + " (current)" : opt.Name;
            string icon = opt.Icon == "" ? "●" : opt.Icon;
            int lineLen = $"> {icon} {label}".Length;
            if (lineLen > maxWidth) maxWidth = lineLen;
        }

        var optionList = new System.Text.StringBuilder();
        for (int i = 0; i < _options.Count; i++)
        {
            var opt = _options[i];
            string cursor = i == _cursor ? ">" : " ";
            string label = opt.IsCurrent ? opt.Name + " (current)" : opt.Name;
            string icon = opt.Icon == "" ? "●" : opt.Icon;
            string line = $"{cursor} {icon} {label}";

            line = i == _cursor
                ? Style.New().Foreground(theme.SelectForeground).Background(theme.SelectBackground).Width(maxWidth).Render(line)
                : Style.New().Foreground(theme.Foreground).Background(theme.Background).Width(maxWidth).Render(line);

            optionList.Append(line).Append('\n');
        }

        string title = Style.New()
            .Foreground(theme.Primary)
            .Background(theme.Background)
            .Bold()
            .Width(maxWidth)
            .Render(_title);

        string helpText = Style.New()
            .Foreground(theme.ForegroundMuted)
            .Background(theme.Background)
            .Width(maxWidth)
            .Render(helpTextStr);

        string content = Layout.JoinVertical(HAlign.Left, title, "", optionList.ToString(), helpText);

        var modalStyle = Style.New()
            .WithBorder(Border.Rounded)
            .BorderForeground(theme.Border)
            .Padding(1, 2)
            .Background(theme.Background);

        var modal = modalStyle.Render(content);

        if (_width > 0 && _height > 0)
            modal = Layout.Place(_width, _height, HAlign.Center, VAlign.Center, modal);

        return modal;
    }
}

File: reference/internal/ui/components/listpicker.go
package components

import (
	"fmt"

	"github.com/Elpulgo/azdo/internal/ui/styles"
	"github.com/charmbracelet/bubbles/key"
	tea "github.com/charmbracelet/bubbletea"
	"github.com/charmbracelet/lipgloss"
)

const minModalWidth = 80

type ListPickerOption struct {
	Name      string
	Icon      string
	IsCurrent bool
}

type ListPickerSelectedMsg struct {
	Value string
}

type ListPicker struct {
	styles     *styles.Styles
	visible    bool
	width      int
	height     int
	title      string
	options    []ListPickerOption
	cursor     int
	showClear  bool
	allowClear bool
}

func NewListPicker(s *styles.Styles) ListPicker {
	return ListPicker{
		styles:  s,
		visible: false,
		cursor:  0,
	}
}

func (lp *ListPicker) SetConfig(title string, options []ListPickerOption, activeValue string, allowClear bool) {
	lp.title = title
	lp.allowClear = allowClear
	lp.options = nil
	lp.cursor = 0

	hasActive := activeValue != "" && allowClear

	if hasActive {
		lp.options = append(lp.options, ListPickerOption{Name: "Clear filter", Icon: "✕", IsCurrent: false})
	}

	offset := 0
	if hasActive {
		offset = 1
	}

	for i, opt := range options {
		opt.IsCurrent = opt.Name == activeValue
		lp.options = append(lp.options, opt)
		if opt.Name == activeValue {
			lp.cursor = i + offset
		}
	}
}

func (lp *ListPicker) Show() {
	lp.visible = true
}

func (lp *ListPicker) Hide() {
	lp.visible = false
}

func (lp ListPicker) IsVisible() bool {
	return lp.visible
}

func (lp *ListPicker) SetSize(width, height int) {
	lp.width = width
	lp.height = height
}

func (lp ListPicker) GetCursor() int {
	return lp.cursor
}

func (lp ListPicker) Update(msg tea.Msg) (ListPicker, tea.Cmd) {
	if !lp.visible {
		return lp, nil
	}

	switch msg := msg.(type) {
	case tea.KeyMsg:
		switch {
		case key.Matches(msg, key.NewBinding(key.WithKeys("esc", "q"))):
			lp.visible = false
			return lp, nil

		case key.Matches(msg, key.NewBinding(key.WithKeys("up", "k"))):
			if lp.cursor > 0 {
				lp.cursor--
			}
			return lp, nil

		case key.Matches(msg, key.NewBinding(key.WithKeys("down", "j"))):
			if lp.cursor < len(lp.options)-1 {
				lp.cursor++
			}
			return lp, nil

		case key.Matches(msg, key.NewBinding(key.WithKeys("enter"))):
			if len(lp.options) == 0 {
				return lp, nil
			}
			selected := lp.options[lp.cursor]
			lp.visible = false
			value := selected.Name
			if selected.Name == "Clear filter" {
				value = ""
			}
			return lp, func() tea.Msg {
				return ListPickerSelectedMsg{Value: value}
			}
		}
	}

	return lp, nil
}

func (lp ListPicker) View() string {
	if !lp.visible {
		return ""
	}

	helpTextStr := "↑/↓: navigate • enter: select • esc/q: cancel"

	maxWidth := minModalWidth
	if len(lp.title) > maxWidth {
		maxWidth = len(lp.title)
	}
	if len(helpTextStr) > maxWidth {
		maxWidth = len(helpTextStr)
	}

	for _, opt := range lp.options {
		label := opt.Name
		if opt.IsCurrent {
			label += " (current)"
		}
		icon := opt.Icon
		if icon == "" {
			icon = "●"
		}
		lineLen := len(fmt.Sprintf("> %s %s", icon, label))
		if lineLen > maxWidth {
			maxWidth = lineLen
		}
	}

	var optionList string
	for i, opt := range lp.options {
		cursor := " "
		if i == lp.cursor {
			cursor = ">"
		}

		label := opt.Name
		if opt.IsCurrent {
			label += " (current)"
		}

		icon := opt.Icon
		if icon == "" {
			icon = "●"
		}

		line := fmt.Sprintf("%s %s %s", cursor, icon, label)

		if i == lp.cursor {
			line = lipgloss.NewStyle().
				Foreground(lp.styles.Theme.GetSelectForeground()).
				Background(lp.styles.Theme.GetSelectBackground()).
				Width(maxWidth).
				Render(line)
		} else {
			line = lipgloss.NewStyle().
				Foreground(lp.styles.Theme.GetForeground()).
				Background(lp.styles.Theme.GetBackground()).
				Width(maxWidth).
				Render(line)
		}

		optionList += line + "\n"
	}

	title := lipgloss.NewStyle().
		Foreground(lp.styles.Theme.GetPrimary()).
		Background(lp.styles.Theme.GetBackground()).
		Bold(true).
		Width(maxWidth).
		Render(lp.title)

	helpText := lipgloss.NewStyle().
		Foreground(lp.styles.Theme.GetForegroundMuted()).
		Background(lp.styles.Theme.GetBackground()).
		Width(maxWidth).
		Render(helpTextStr)

	content := lipgloss.JoinVertical(
		lipgloss.Left,
		title,
		"",
		optionList,
		helpText,
	)

	modalStyle := lipgloss.NewStyle().
		Border(lipgloss.RoundedBorder()).
		BorderForeground(lp.styles.Theme.GetBorder()).
		Padding(1, 2).
		Background(lp.styles.Theme.GetBackground())

	modal := modalStyle.Render(content)

	if lp.width > 0 && lp.height > 0 {
		modal = lipgloss.Place(
			lp.width,
			lp.height,
			lipgloss.Center,
			lipgloss.Center,
			modal,
		)
	}

	return modal
}
